velox/exec/tests/utils/TpchQueryBuilder.cpp

```diff
@@ -612,6 +614,69 @@ TpchPlan TpchQueryBuilder::getQ13Plan() const {
   return context;
 }
 
+TpchPlan TpchQueryBuilder::getQ14Plan() const {
+  std::vector<std::string> lineitemColumns = {
+      "l_partkey", "l_extendedprice", "l_discount", "l_shipdate"};
+  std::vector<std::string> partColumns = {"p_partkey", "p_type"};
+
+  auto lineitemSelectedRowType = getRowType(kLineitem, lineitemColumns);
+  const auto& lineitemFileColumns = getFileColumnNames(kLineitem);
+  auto partSelectedRowType = getRowType(kPart, partColumns);
+  const auto& partFileColumns = getFileColumnNames(kPart);
+
+  std::string shipDate = "l_shipdate";
+  std::string shipDateFilter = formatDateFilter(
+      shipDate, lineitemSelectedRowType, "'1995-09-01'", "'1995-09-30'");
+
+  auto planNodeIdGenerator = std::make_shared<PlanNodeIdGenerator>();
+  core::PlanNodeId lineitemScanNodeId;
+  core::PlanNodeId partScanNodeId;
+
+  auto part = PlanBuilder(planNodeIdGenerator)
+                  .tableScan(kPart, partSelectedRowType, partFileColumns)
+                  .capturePlanNodeId(partScanNodeId)
+                  .planNode();
+
+  auto plan =
+      PlanBuilder(planNodeIdGenerator)
+          .tableScan(
+              kLineitem,
+              lineitemSelectedRowType,
+              lineitemFileColumns,
+              {},
+              shipDateFilter)
+          .capturePlanNodeId(lineitemScanNodeId)
+          .project(
+              {"l_extendedprice * (1.0 - l_discount) as part_revenue",
+               "l_shipdate",
+               "l_partkey"})
+          .hashJoin(
+              {"l_partkey"},
+              {"p_partkey"},
+              part,
+              "",
+              {"part_revenue", "p_type"})
+          .project(
+              {"(CASE WHEN (p_type LIKE 'PROMO%') THEN part_revenue ELSE 0.0 END) as filter_revenue",
+               "part_revenue"})
+          .partialAggregation(
+              {},
+              {"sum(part_revenue) as total_revenue",
```

Comment:
can we embed the `CASE` expression inside `sum` here?

Reply:
Unsure if HashAgg supports non-field or non-constant expressions here. I think this will crash in Operator::exprToChannel(...) But might be worth checking.

Reply:
Inside `AggregationTest.cpp`, I see some simple expressions like `sum(c0 + c1)`. I am curious about `CASE` now.

Reply:
Thanks for the suggestion but using CASE in sum gives the following error:
Error Source: USER
Error Code: INVALID_ARGUMENT
Reason: Aggregate function doesn't exist: if.

Reply:
Can you share the backtrace?

Reply:
__cxa_throw 0x00007ff80836ec58
facebook::velox::detail::veloxCheckFail<…>(const facebook::velox::detail::VeloxCheckFailArgs &, const std::string &) Exceptions.h:74
facebook::velox::exec::test::throwAggregateFunctionDoesntExist(const std::string &) PlanBuilder.cpp:284
facebook::velox::exec::test::resolveAggregateType(const std::string &, facebook::velox::core::AggregationNode::Step, const std::vector<…> &, bool) PlanBuilder.cpp:354
facebook::velox::exec::test::AggregateTypeResolver::resolveType(const std::vector<…> &, const std::shared_ptr<…> &, bool) const PlanBuilder.cpp:395
<lambda>::operator()<…>(const auto &, const auto &, bool) const PlanBuilder.cpp:364
std::__invoke<…>(auto &&, const std::vector<…> &, const std::shared_ptr<…> &, bool &&) type_traits:3918
std::__invoke_void_return_wrapper::__call<…>(<lambda> &, const std::vector<…> &, const std::shared_ptr<…> &, bool &&) invoke.h:30
std::__function::__alloc_func::operator()(const std::vector<…> &, const std::shared_ptr<…> &, bool &&) function.h:178
std::__function::__func::operator()(const std::vector<…> &, const std::shared_ptr<…> &, bool &&) function.h:352
std::__function::__value_func::operator()(const std::vector<…> &, const std::shared_ptr<…> &, bool &&) const function.h:505
std::function::operator()(const std::vector<…> &, const std::shared_ptr<…> &, bool) const function.h:1182
facebook::velox::core::resolveTypeImpl(std::vector<…>, const std::shared_ptr<…> &, bool) Expressions.cpp:51
facebook::velox::core::createWithImplicitCast(std::shared_ptr<…>, std::vector<…>) Expressions.cpp:171
facebook::velox::core::Expressions::inferTypes(const std::shared_ptr<…> &, const std::shared_ptr<…> &, facebook::velox::memory::MemoryPool *) Expressions.cpp:223
facebook::velox::core::Expressions::inferTypes(const std::shared_ptr<…> &, const std::shared_ptr<…> &, facebook::velox::memory::MemoryPool *) Expressions.cpp:207
facebook::velox::exec::test::PlanBuilder::inferTypes(const std::shared_ptr<…> &) PlanBuilder.cpp:1096
facebook::velox::exec::test::PlanBuilder::createAggregateExpressionsAndNames(const std::vector<…> &, facebook::velox::core::AggregationNode::Step, const std::vector<…> &) PlanBuilder.cpp:529
facebook::velox::exec::test::PlanBuilder::aggregation(const std::vector<…> &, const std::vector<…> &, const std::vector<…> &, const std::vector<…> &, facebook::velox::core::AggregationNode::Step, bool, const std::vector<…> &) PlanBuilder.cpp:572
facebook::velox::exec::test::PlanBuilder::partialAggregation(const std::vector<…> &, const std::vector<…> &, const std::vector<…> &) PlanBuilder.h:280
facebook::velox::exec::test::TpchQueryBuilder::getQ14Plan() const TpchQueryBuilder.cpp:662
facebook::velox::exec::test::TpchQueryBuilder::getQueryPlan(int) const TpchQueryBuilder.cpp:123
ParquetTpchTest::assertQuery(int, int, int, std::optional<…>) const ParquetTpchTest.cpp:120
ParquetTpchTest_Q14_Test::TestBody() ParquetTpchTest.cpp:200
testing::internal::HandleSehExceptionsInMethodIfSupported<…>(testing::Test *, void (testing::Test::*)(), const char *) gtest.cc:2607
testing::internal::HandleExceptionsInMethodIfSupported<…>(testing::Test *, void (testing::Test::*)(), const char *) gtest.cc:2643
testing::Test::Run() gtest.cc:2682
testing::TestInfo::Run() gtest.cc:2861
testing::TestSuite::Run() gtest.cc:3015
testing::internal::UnitTestImpl::RunAllTests() gtest.cc:5855
testing::internal::HandleSehExceptionsInMethodIfSupported<…>(testing::internal::UnitTestImpl *, bool (testing::internal::UnitTestImpl::*)(), const char *) gtest.cc:2607
testing::internal::HandleExceptionsInMethodIfSupported<…>(testing::internal::UnitTestImpl *, bool (testing::internal::UnitTestImpl::*)(), const char *) gtest.cc:2643
testing::UnitTest::Run() gtest.cc:5438
RUN_ALL_TESTS() gtest.h:2490
main ParquetTpchTest.cpp:210
start 0x000000013991451e

q14 stack when case is used in aggregate function sum

Reply:
seems to be a limitation of the plan builder. But an extra project step should be okay for now.